cadence/src/builder.rs

```diff
@@ -195,11 +236,49 @@ impl<'a> MetricFormatter<'a> {
         Self::TAG_PREFIX.len() + self.kv_size + self.tags.len() - 1
     }
 
+    fn timestamp_size_hint(&self) -> usize {
+        if let Some(_timestamp) = self.timestamp {
+            /* |T */
+            2 + /* timestamp */ 10
+        } else {
+            0
+        }
+    }
+
+    fn sampling_rate_size_hint(&self) -> usize {
+        if let Some(_rate) = self.sampling_rate {
+            /* |@ */
+            2 + /* rate */ 10
```

Comment:
nit: might be good to comment why 10 is a good number here.

Reply:
done